### DISCUSSION TITLE
Docker + Ubuntu 22.04: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.

### DISCUSSION THREAD

**Discussion by jaydrogers** [oerdnj/deb.sury.org#1839] [Q&A] [answered]
**Frequently asked questions**

- [x] I have read [Frequently Asked Questions](https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions)

## Background
I would like to use the "ondrej/php" repository inside of my container image, so installing `software-properties` common will greatly inflate the size of the Docker image if I want `add-apt-repository`.

I am currently adding the repository via this method:
```bash
echo "deb http://ppa.launchpad.net/ondrej/php/ubuntu $(bash -c '. /etc/os-release;echo $VERSION_CODENAME') main" > /etc/apt/sources.list.d/ondrej-php.list
```

## Problem
The above command works, but it is throwing this deprecation error:
```txt
http://ppa.launchpad.net/ondrej/php/ubuntu/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
```

## Attempted solution
[I saw how Docker is adding keys](https://docs.docker.com/engine/install/ubuntu/) and [this article](https://itsfoss.com/apt-key-deprecated/), so I attempted to adapt these methods from the base Ubuntu 22.04 image:

```bash
apt update
apt-get install -y --no-install-recommends apt-transport-https gnupg2 curl ca-certificates unzip
curl -fsSL https://ppa.launchpadcontent.net/ondrej/php/ubuntu/dists/$(bash -c '. /etc/os-release;echo $VERSION_CODENAME')/Release.gpg | gpg --dearmor -o /usr/share/keyrings/ondrej-php.gpg
echo "deb [signed-by=/usr/share/keyrings/ondrej-php.gpg] http://ppa.launchpad.net/ondrej/php/ubuntu $(bash -c '. /etc/os-release;echo $VERSION_CODENAME') main" | tee /etc/apt/sources.list.d/ppa-php.list > /dev/null
```

### Error I am getting with attempted solution
<img width="2202" alt="image" src="https://user-images.githubusercontent.com/3174134/192053121-ef94ce53-0e41-42b7-8a0e-86e059d61bea.png">

**Error:**
```
GPG error: http://ppa.launchpad.net/ondrej/php/ubuntu jammy InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 4F4EA0AAE5267A6C
```

**Contents of `/etc/apt/sources.list.d/ppa-php.list`:**
```txt
deb [signed-by=/usr/share/keyrings/ondrej-php.gpg] http://ppa.launchpad.net/ondrej/php/ubuntu jammy main
``` 

#### Run a fresh docker container
```
docker run --rm -it ubuntu:22.04 bash
```
#### Paste in my configuration commands
```
apt update
apt-get install -y --no-install-recommends apt-transport-https gnupg2 curl ca-certificates unzip
curl -fsSL https://ppa.launchpadcontent.net/ondrej/php/ubuntu/dists/$(bash -c '. /etc/os-release;echo $VERSION_CODENAME')/Release.gpg | gpg --dearmor -o /usr/share/keyrings/ondrej-php.gpg
echo "deb [signed-by=/usr/share/keyrings/ondrej-php.gpg] http://ppa.launchpad.net/ondrej/php/ubuntu $(bash -c '. /etc/os-release;echo $VERSION_CODENAME') main" | tee /etc/apt/sources.list.d/ppa-php.list > /dev/null
```
### Try updating the packages in the temporary container
```
apt update
```

## Question
1. Is there something that I am missing? Has anyone else run into this issue before?

I am _very grateful_ for the years of hard work you've been doing to maintain this repository. If there's anything else that I can clarify, let me know. I would be grateful for any insight here!

**Comment by oerdnj** [comment]
> I am currently adding the repository via this method

Care to explain why?

It seems like you invented a more complicated method that doesn’t work instead of just using a standard method for adding PPA to the system, so there must be some thought behind this. Right?

**Reply by jaydrogers** [reply]
Thanks for getting back!

I understand this method is more complicated. The reason is because this is within a container environment.

To add the `add-apt-repository` command, I need `software-properties-common`.

This would increase my image size by well over 100MB (which is almost double of the current image size).

I’m hoping I can achieve the same results without adding a lot of packages.

If you have any thoughts on how to achieve this, I would love to learn more. Thanks!

**Reply by oerdnj** [reply]
`software-properties-common` doesn’t have to stay installed after you are done. You would be better asking the basic questions at AskUbuntu though.

**Comment by rfay** [comment]
All of the traditional key installation methods are getting this complaint on Ubuntu 22.04.

**Comment by jaydrogers** [comment, marked as answer]
# ✅ Solution
Using [Docker Multi-stage builds](https://docs.docker.com/build/building/multi-stage/) solved my issue. This allowed me to:

1. Follow the supported methods of Ondřej
2. Keep my images small.

I was able to take the original method from 227 MB → 181MB (both sizes are listed in their _uncompressed_ size)

Here is the source code that worked for me without having any errors appear when running `apt-get update`:

```Dockerfile
FROM ubuntu:22.04 as repo-config

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update \
    && apt-get install -y --no-install-recommends gnupg2 software-properties-common \
    && add-apt-repository -y ppa:ondrej/php

######################################################################################
FROM ubuntu:22.04

ENV DEBIAN_FRONTEND=noninteractive

COPY --from=repo-config /etc/apt/sources.list.d/ /etc/apt/sources.list.d/
COPY --from=repo-config /etc/apt/trusted.gpg.d/ondrej-ubuntu-php.gpg /etc/apt/trusted.gpg.d/ondrej-ubuntu-php.gpg

RUN apt-get update \
	&& apt-get -y --no-install-recommends install php8.1-cli \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* /usr/share/doc/*
```
